src/metadata.cpp

```diff
@@ -0,0 +1,260 @@
+    *d = *other.d;
+}
+
+// ---------------------------------------------------------------------------
+
+Extent::~Extent() = default;
+
+// ---------------------------------------------------------------------------
+
+const optional<std::string> &Extent::description() const {
+    return d->description;
+}
+
+// ---------------------------------------------------------------------------
+
+//! @cond Doxygen_Suppress
+struct Identifier::Private {
+    optional<Citation> authority{};
+    std::string code{};
+    optional<std::string> codeSpace{};
+    optional<std::string> version{};
+    optional<std::string> description{};
+};
+//! @endcond
+
+// ---------------------------------------------------------------------------
+
+Identifier::Identifier(const std::string &codeIn)
+    : d(internal::make_unique<Private>()) {
+    d->code = codeIn;
+}
+
+// ---------------------------------------------------------------------------
+
+Identifier::Identifier(const Identifier &other)
+    : d(internal::make_unique<Private>(*(other.d))) {}
+
+// ---------------------------------------------------------------------------
+
+Identifier::~Identifier() = default;
+
+// ---------------------------------------------------------------------------
+
+IdentifierNNPtr Identifier::create(const std::string &codeIn,
+                                   const PropertyMap &properties) {
+    auto id(Identifier::nn_make_shared<Identifier>(codeIn));
+    id->setProperties(properties);
+    return id;
+}
+
+// ---------------------------------------------------------------------------
+
+Identifier &Identifier::operator=(const Identifier &other) {
+    if (this != &other) {
+        *d = *other.d;
+    }
+    return *this;
+}
+
+// ---------------------------------------------------------------------------
+
+const optional<Citation> &Identifier::authority() const { return d->authority; }
+
+// ---------------------------------------------------------------------------
+
+const std::string &Identifier::code() const { return d->code; }
+
+// ---------------------------------------------------------------------------
+
+const optional<std::string> &Identifier::codeSpace() const {
+    return d->codeSpace;
+}
+
+// ---------------------------------------------------------------------------
+
+const optional<std::string> &Identifier::version() const { return d->version; }
+
+// ---------------------------------------------------------------------------
+
+const optional<std::string> &Identifier::description() const {
+    return d->description;
+}
+
+// ---------------------------------------------------------------------------
+
+void Identifier::setProperties(
+    const PropertyMap &properties) // throw(InvalidValueTypeException)
+{
+    {
+        auto oIter = properties.find(AUTHORITY_KEY);
+        if (oIter != properties.end()) {
+            auto genVal =
+                util::nn_dynamic_pointer_cast<BoxedValue>(oIter->second);
+            if (genVal) {
```

Comment:
This shorter version of if-statement with single variable declaration and initialization is C++ idiomatic, no compiler should warn:

```
if (auto genVal = util::nn_dynamic_pointer_cast<BoxedValue>(oIter->second))
{
  assert(genVal);
}
else
{
  assert(!genVal);
}
```